Subsample fastq 38

FlicAnderson opened the pull request:
Finally ready to merge this rather old branch in; have checked it works, and have adjusted path for the example 'how to use'. 

Merged latest develop into it, and ran all tests (all passing). 

Used this on an [example-datasets#/10](https://github.com/riboviz/example-datasets/issues/10) sample.

FlicAnderson commented:
Just realised that I don't understand the relationship between `riboviz/subsample_bioseqfile.py` and `riboviz/tools/subsample_bioseqfile.py` and this is probably important.  I hadn't noticed the one in `riboviz/tools` so haven't changed that one yet.

@mikej888 can you explain a little about how these are set up (ie why it seems that for certain items in `riboviz/` there's a corresponding item in `riboviz/tools/` and let me know what I need to do to make sure I haven't broken some capability you spent ages setting up? :grin:

mikej888 commented:
Scripts in `riboviz/tools/<NAME>.py` are those scripts intended to be run from the command-line. They include the `__main__` block and functions to set up the command-line parameters, parse these, extract the values then invoke functions in `riboviz/<NAME>.py`. It is to decouple the user interface from the actual functionality.

FlicAnderson commented:
In which case, I need to go back and make sure the version in `riboviz/tools/` and the other version match up ok, so in which case, this probably shouldn't be merged before I try and do that, I think?

mikej888 commented:
Yes, refactoring it to be consistent with the other command-line tools would be good. Whenever I need need to write a new command-line tool I usually copy one from riboviz/tools/ !

FlicAnderson commented:
@mikej888 As of commit `3c6d91` I've attempted to make `riboviz/subsample_bioseqfile.py` and  `riboviz/tools/subsample_bioseqfile.py` in branch `subsample-fastq-38` more consistent with other riboviz functions & modules, but I'm not 100% sure it's quite right. 

I have checked that it runs, and it does seem to do what I expect, although I did test on `.fastq.gz` files so couldn't use `wc` to quickly check number of lines of input and output files to confirm, and have to sign off for tonight without having time to unzip some .fastq.gz files to check the inputs/outputs with fasta files. 

I'm not confident that the function actually does anything useful with the `overwrite` parameter if it's there and used...  :confused: 

I'm going to try and do more testing of this tomorrow, and also attempt to run some of the python style checkers mentioned in the developer documentation. 

I've merged `develop` into `subsample-fastq-38` today also, so it should hopefully merge nicely once it's complete.

If you have any time to skim your eye over these files and see if you can find any obvious fails / inconsistencies ahead of the dev meeting, that would be great, but no worries if not.

FlicAnderson commented:
Haven't had time to do any more testing/investigating on this code yet, that will probably happen tomorrow/Friday this week instead. 

It would be good to wrap this up and get it merged in out of the way, as this code would be particularly helpful for @3mma-mack and @swinterbourne as they work on new datasets for their projects this term.

FlicAnderson commented:
Ok, so. 
I've run & made edits according to `2to3`, and `pycodestyle`. Those are resolved. 

I've run `pylint` and resolved everything it complained about except a few crucial bits that I don't seem to be able to fix.

```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/tools/subsample_bioseqfile.py 
************* Module riboviz.tools.subsample_bioseqfile
riboviz/tools/subsample_bioseqfile.py:87:4: W0101: Unreachable code (unreachable)
------------------------------------------------------------------
Your code has been rated at 9.67/10 (previous run: 9.67/10, +0.00)

(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/subsample_bioseqfile.py 
************* Module riboviz.subsample_bioseqfile
riboviz/subsample_bioseqfile.py:36:2: W0511: TODO there needs to be some conditional with 'overwrite' here (fixme)
riboviz/subsample_bioseqfile.py:10:0: R0913: Too many arguments (6/5) (too-many-arguments)
riboviz/subsample_bioseqfile.py:11:47: W0613: Unused argument 'overwrite' (unused-argument)
------------------------------------------------------------------
Your code has been rated at 8.70/10 (previous run: 9.66/10, -0.96)
```

I can't get the hang of what's going on with the "unreachable code":
```python
    if os.path.exists(options.seqfileout) and not options.overwrite:
        raise ValueError(
            "output file {} already exists, use '-overwrite' to replace"
            .format(options.seqfileout))
```
If the output file path already exists and the overwrite argument is not set to true (ie do not overwrite), then as far as I can see, shouldn't this code return the error "output file {} already exists, use '-overwrite' to replace"?

I was wondering if this was some larger logic puzzle where because os.path.exists() is TRUE and "not options.overwrite" would be the same as "not TRUE, therefore FALSE" and therefore the whole AND statement doesn't get run because TRUE + FALSE = FALSE? But then I realised I still didn't know how to fix it nicely if that was the case, so I'm calling in the experts!

I also keep going backwards and forwards about what we might want to happen if we have NOT set overwrite to true but the file exists.  Initially I tried some nested if statements which would try appending the extra lines into the existing file? But I think this is probably a 2-line fix by someone who knows more python :grinning:  

Requesting backup for a code review to help finish and resolve this issue & pre-emptive PR: @mikej888  or @ewallace

mikej888 commented:
For:
```
riboviz/tools/subsample_bioseqfile.py:87:4: W0101: Unreachable code (unreachable)
```
this relates to my comment above on the code.

> These lines are after return and will never be called. The existence of files and all other checks should be done in riboviz.subsample_bioseqfile.subsample_bioseqfile(...).

i.e. the block
```
    ...
    return options

    # files exist, overwrite output?
    if os.path.exists(options.seqfileout) and not options.overwrite:
        raise ValueError(
            "output file {} already exists, use '-overwrite' to replace"
            .format(options.seqfileout))
    if not os.path.exists(options.seqfilein):
        raise ValueError(
            "input file {} doesn't exist".format(options.seqfilein))
```

is within `def parse_command_line_options():` but is after `return` so it is unreachable.  The existence of files and all other checks should be done in `riboviz.subsample_bioseqfile.subsample_bioseqfile(...)`. Moving the above into that function then replacing `options<N>` with `<N>` where `<N>` is one of the arguments to `subsample_bioseqfile` should both fix this and the warning:

```
riboviz/subsample_bioseqfile.py:11:47: W0613: Unused argument 'overwrite' (unused-argument)
```

For
```
riboviz/subsample_bioseqfile.py:10:0: R0913: Too many arguments (6/5) (too-many-arguments)
```
that can be ignored. pylint usually complains if you have 6 or more arguments to a function! It complains with good reason but it'd be effort to refactor the code to avoid this e.g. by introducing additional data structures. I tend to ignore it.

FlicAnderson commented:
Thanks @mikej888 - I hadn't clocked that putting code after `return` doesn't do much good, but as soon as you mentioned it, it made sense!  
Likewise re-positioning the file-checking code block into the main script also made much more sense, as it's 'functional' code, rather than the 'meta' code in the `tools/` version which I probably should think of as like a 'wrapper'. 

I've made the edits you've suggested and removed the TODO.

After last batch of fixing:
```
# check file to pretend-overwrite already exists: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ ls vignette/tmp/
subsamplefile.fastq.gz  WT3AT  WTnone

# run command WITHOUT overwrite:
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsamplefile.fastq.gz -p 0.001
Created by: RiboViz
Date: 2021-01-27 15:08:30.639181
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit cdea2af685003e224a3b1a75b4a9f8a0ad7615fb date 2021-01-26 15:28:26+00:00

Traceback (most recent call last):
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 108, in <module>
    invoke_subsample_bioseqfile()
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 104, in invoke_subsample_bioseqfile
    verbose)
  File "/home/fanders6/riboviz/riboviz/riboviz/subsample_bioseqfile.py", line 40, in subsample_bioseqfile
    .format(seqfileout))
ValueError: output file vignette/tmp/subsamplefile.fastq.gz already exists, use '-overwrite' to replace
# NICE!

# Run WITH overwrite = TRUE
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsamplefile.fastq.gz -p 0.001 -f True
Created by: RiboViz
Date: 2021-01-27 15:22:25.528274
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit cdea2af685003e224a3b1a75b4a9f8a0ad7615fb date 2021-01-26 15:28:26+00:00
read 100000
read 200000
read 300000
read 400000
read 500000
read 600000
read 700000
read 800000
read 900000
subsampling complete; read 963571 records from vignette/input/SRR1042855_s1mi.fastq.gz,         wrote 988 records to vignette/tmp/subsamplefile.fastq.gz
```
So the script now DOES warn you if you've specified an existing output file. But it also happily overwrites it once you run with overwrite = True.  This is what we were hoping for. :+1: 

* checked pylint outputs now:
```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/subsample_bioseqfile.py 
************* Module riboviz.subsample_bioseqfile
riboviz/subsample_bioseqfile.py:10:0: R0913: Too many arguments (6/5) (too-many-arguments)
------------------------------------------------------------------
Your code has been rated at 9.63/10 (previous run: 9.26/10, +0.37)

(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/tools/subsample_bioseqfile.py 
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
This is good. Leaves only the too-many-arguments complaint @mikej888 mentioned is ignoreable :)

Merged in latest develop & ran vignette & tests (using latest regression test dataset)
```
# run nextflow vignette
(riboviz) fanders6@ceres:~/riboviz/riboviz$ nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml
N E X T F L O W  ~  version 20.07.1
Launching `prep_riboviz.nf` [silly_colden] - revision: 16531346ed
No such sample file (NotHere): example_missing_file.fastq.gz
executor >  local (24)
executor >  local (24)
[49/f8b7fa] process > buildIndicesrRNA (yeast_rRNA)   [100%] 1 of 1 ✔
[f3/4f8142] process > buildIndicesORF (YAL_CDS_w_250) [100%] 1 of 1 ✔
[6a/f20d98] process > cutAdapters (WT3AT)             [100%] 2 of 2 ✔
[-        ] process > extractUmis                     -
[-        ] process > cutAdaptersMultiplex            -
[-        ] process > extractUmisMultiplex            -
[-        ] process > demultiplex                     -
[60/9221d2] process > hisat2rRNA (WT3AT)              [100%] 2 of 2 ✔
[4d/a66287] process > hisat2ORF (WT3AT)               [100%] 2 of 2 ✔
[e6/66507c] process > trim5pMismatches (WT3AT)        [100%] 2 of 2 ✔
[76/73c6fb] process > samViewSort (WT3AT)             [100%] 2 of 2 ✔
[-        ] process > groupUmisPreDedup               -
[-        ] process > dedupUmis                       -
[-        ] process > groupUmisPostDedup              -
[71/4cbf86] process > outputBams (WT3AT)              [100%] 2 of 2 ✔
[cb/e90c37] process > makeBedgraphs (WT3AT)           [100%] 2 of 2 ✔
[51/7cfc6f] process > bamToH5 (WT3AT)                 [100%] 2 of 2 ✔
[ff/4f0dc6] process > generateStatsFigs (WT3AT)       [100%] 2 of 2 ✔
[2c/b99194] process > renameTpms (WT3AT)              [100%] 2 of 2 ✔
[e3/b004c8] process > collateTpms (WTnone, WT3AT)     [100%] 1 of 1 ✔
[44/8c7a38] process > countReads                      [100%] 1 of 1 ✔

Completed at: 27-Jan-2021 15:32:52
Duration    : 1m 50s
CPU hours   : (a few seconds)
Succeeded   : 24
```
Looking good.

* downloaded new regression test dataset: 
https://github.com/riboviz/regression-test-data-20210119
```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/regression-test-data-20210119 --config-file=vignette/vignette_config.yaml --nextflow
========================================== test session starts ==========================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/fanders6/riboviz/riboviz
plugins: cov-2.10.1
collected 88 items                                                                                      

riboviz/test/regression/test_regression.py ssssssssssssssssssssssssssssssssssssss..ssssssssss.... [ 61%]
..................................                                                                [100%]

=========================================== warnings summary ============================================
/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/site-packages/Bio/Alphabet/__init__.py:27
  /home/fanders6/miniconda3/envs/riboviz/lib/python3.7/site-packages/Bio/Alphabet/__init__.py:27: PendingDeprecationWarning: We intend to remove or replace Bio.Alphabet in 2020, ideally avoid using it explicitly in your code. Please get in touch if you will be adversely affected by this. https://github.com/biopython/biopython/issues/2046
    PendingDeprecationWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
========================= 40 passed, 48 skipped, 1 warning in 120.91s (0:02:00) =========================
```
Nice. 

Run other tests:
```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pytest --ignore-glob="*regression*"
========================================== test session starts ==========================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/fanders6/riboviz/riboviz
plugins: cov-2.10.1
collected 286 items                                                                                     

riboviz/test/test_barcodes_umis.py .................                                              [  5%]
riboviz/test/test_check_fasta_gff.py .......................                                      [ 13%]
riboviz/test/test_demultiplex_fastq.py ..................                                         [ 20%]
riboviz/test/test_fastq.py ........................                                               [ 28%]
riboviz/test/test_get_cds_codons.py ........................................                      [ 42%]
riboviz/test/test_process_utils.py ........................                                       [ 51%]
riboviz/test/test_sam_bam.py ................                                                     [ 56%]
riboviz/test/test_sample_sheet.py ..........                                                      [ 60%]
riboviz/test/test_trim_5p_mismatch.py ............                                                [ 64%]
riboviz/test/test_upgrade_config.py .....                                                         [ 66%]
riboviz/test/test_utils.py ....................                                                   [ 73%]
riboviz/test/nextflow/test_nextflow_errors.py ............................................        [ 88%]
riboviz/test/tools/test_prep_riboviz.py ............                                              [ 92%]
riboviz/test/tools/test_prep_riboviz_simdata_multiplex.py ..............                          [ 97%]
riboviz/test/tools/test_prep_riboviz_simdata_umi.py .......                                       [100%]

=========================================== warnings summary ============================================
/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/site-packages/Bio/Alphabet/__init__.py:27
  /home/fanders6/miniconda3/envs/riboviz/lib/python3.7/site-packages/Bio/Alphabet/__init__.py:27: PendingDeprecationWarning: We intend to remove or replace Bio.Alphabet in 2020, ideally avoid using it explicitly in your code. Please get in touch if you will be adversely affected by this. https://github.com/biopython/biopython/issues/2046
    PendingDeprecationWarning,

riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_use_feature_name_true
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_cds_format
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_exclude_stop_codons
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_use_feature_name_true
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_cds_format
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_exclude_stop_codons
  /home/fanders6/riboviz/riboviz/riboviz/get_cds_codons.py:171: UserWarning: 'YAL003CMissingGene_mRNA not in /home/fanders6/riboviz/riboviz/riboviz/test/data/test_get_cds_codons.fasta.'
    warnings.warn(str(e))

riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_use_feature_name_true
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_cds_format
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_from_fasta_exclude_stop_codons
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_use_feature_name_true
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_cds_format
riboviz/test/test_get_cds_codons.py::test_get_cds_codons_file_exclude_stop_codons
  /home/fanders6/riboviz/riboviz/riboviz/get_cds_codons.py:174: UserWarning: Feature YAL007CBadLengthCDS_mRNA (ATGGCCTGTTA) has length not divisible by 3
    warnings.warn(str(e))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================= 286 passed, 17 warnings in 246.47s (0:04:06) ==============================
```
Nice. 

I think this should finally be ready to merge!

kavousan commented:
@mikej888 made a suggestion to use "with"; @FlicAnderson to take a look.

FlicAnderson commented on riboviz/tools/subsample_bioseqfile.py:
```diff
@@ -7,40 +7,32 @@
 Usage::
 
     python -m riboviz.tools.subsample_bioseqfile [-h]
-        -i INPUT_FILE -o OUTPUT_FILE
```
Nice tip, thanks!

FlicAnderson commented:
I've made the changes you've suggested @mikej888 but I'm curious how to fix a new error with the code: 

When I run it, I get "wrote 0 records to", and it doesn't seem to write any lines in now. 
I've clearly overlooked something daft? 

```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.0001
Created by: RiboViz
Date: 2021-02-10 15:54:08.943959
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit d9cd7eea0572fc1239266e7be671c6f1e7079786 date 2021-02-10 15:50:47+00:00

read 100000
read 200000
read 300000
read 400000
read 500000
read 600000
read 700000
read 800000
read 900000
subsampling complete; read 963571 records from vignette/input/SRR1042855_s1mi.fastq.gz, wrote 0 records   to vignette/tmp/subsample.fastq.gz


(riboviz) fanders6@ceres:~/riboviz/riboviz$ gunzip vignette/tmp/subsample.fastq.gz 

(riboviz) fanders6@ceres:~/riboviz/riboviz$ wc vignette/tmp/subsample.fastq 
0 0 0 vignette/tmp/subsample.fastq
```
Where did I go wrong **this time**? :grin:

mikej888 commented:
According to https://docs.python.org/3/library/random.html, `random()` returns values between 0..1. The value of `prob`, provided via the command-line, is `0.0001` so `if random() < prob` will always be false. Provide a value between 0 and 1.

On a related note, to guarantee you get the same samples every time, for a given `p`, initialise the random seed. Replace:
```
from random import random
```
with:
```
import random
```
Add, before the `with` block:
```
random.seed(1)
```
Replace:
```
if random() < prob:
```
with:
```
if random.random() < prob:
```

mikej888 commented:
I think this:
```
         if row_count % 100000 == 0:
                print(("read {rowcount}".format(rowcount=row_count)))
                if random() < prob:
                    row_count_out += 1
                    if verbose:
                        print((record.id))
                    SeqIO.write(record, out_handle, filetype)
```
should be:
```
        if row_count % 100000 == 0:
            print(("read {rowcount}".format(rowcount=row_count)))
        if random() < prob:
            row_count_out += 1
            if verbose:
                print((record.id))
            SeqIO.write(record, out_handle, filetype)
```

FlicAnderson commented:
Added set seed option (I used 42), so this gives the same number of results each time now. 

Also fixed the indentation error and made an indentation tweak that removed a gap in the print statement at the end.

Code now runs, gives what is expected, and passes all linter/style tools without anything significant (I think). 

Running the code: 
```
# no overwrite necessary: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.001
Created by: RiboViz
Date: 2021-02-10 18:13:54.755938
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit c18d016dcdc7d3c23bfbcb6dae4aee1087904740 date 2021-02-10 18:03:01+00:00

read 100000
read 200000
read 300000
read 400000
read 500000
read 600000
read 700000
read 800000
read 900000
subsampling complete; read 963571 records from vignette/input/SRR1042855_s1mi.fastq.gz, wrote 921 records to vignette/tmp/subsample.fastq.gz

(riboviz) fanders6@ceres:~/riboviz/riboviz$ gunzip -k vignette/tmp/subsample.fastq.gz 
# unzip & keep original

(riboviz) fanders6@ceres:~/riboviz/riboviz$ wc vignette/tmp/subsample.fastq
  3684   5526 145306 vignette/tmp/subsample.fastq
# 3684 divided by 4 (lines of fastq file per 1x read): 921
# GOOD :)


# with overwrite required, file exists already:
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.01
Created by: RiboViz
Date: 2021-02-10 18:17:34.308764
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit c18d016dcdc7d3c23bfbcb6dae4aee1087904740 date 2021-02-10 18:03:01+00:00

Traceback (most recent call last):
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 109, in <module>
    invoke_subsample_bioseqfile()
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 105, in invoke_subsample_bioseqfile
    verbose)
  File "/home/fanders6/riboviz/riboviz/riboviz/subsample_bioseqfile.py", line 41, in subsample_bioseqfile
    .format(seqfileout))
ValueError: output file vignette/tmp/subsample.fastq.gz already exists, use '-overwrite' to replace
# This is expected, GOOD. 

# re-run WITH overwrite parameter
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.01 -f True
Created by: RiboViz
Date: 2021-02-10 18:11:01.463546
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit c18d016dcdc7d3c23bfbcb6dae4aee1087904740 date 2021-02-10 18:03:01+00:00
read 100000
read 200000
read 300000
read 400000
read 500000
read 600000
read 700000
read 800000
read 900000
subsampling complete; read 963571 records from vignette/input/SRR1042855_s1mi.fastq.gz, wrote 9624 records to vignette/tmp/subsample.fastq.gz

(riboviz) fanders6@ceres:~/riboviz/riboviz$ gunzip -k vignette/tmp/subsample.fastq.gz 
# unzip & keep original

(riboviz) fanders6@ceres:~/riboviz/riboviz$ wc vignette/tmp/subsample.fastq 
  38496   57744 1518651 vignette/tmp/subsample.fastq
# 38496 divided by 4 (lines of fastq file per 1x read): 9624
# GOOD :)

```

pylint output: 
```
# riboviz/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/subsample_bioseqfile.py 
************* Module riboviz.subsample_bioseqfile
riboviz/subsample_bioseqfile.py:10:0: R0913: Too many arguments (6/5) (too-many-arguments)
riboviz/subsample_bioseqfile.py:10:0: R0914: Too many local variables (16/15) (too-many-locals)
------------------------------------------------------------------
Your code has been rated at 9.38/10 (previous run: 9.38/10, +0.00)

# riboviz/tools/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/tools/subsample_bioseqfile.py 
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

pycodestyle output: 
```
# riboviz/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pycodestyle riboviz/subsample_bioseqfile.py 
riboviz/subsample_bioseqfile.py:75:11: E124 closing bracket does not match visual indentation

# riboviz/tools/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pycodestyle riboviz/tools/subsample_bioseqfile.py 
```

2to3 output: 
```
# riboviz/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ 2to3 riboviz/subsample_bioseqfile.py 
RefactoringTool: Skipping optional fixer: buffer
RefactoringTool: Skipping optional fixer: idioms
RefactoringTool: Skipping optional fixer: set_literal
RefactoringTool: Skipping optional fixer: ws_comma
RefactoringTool: No changes to riboviz/subsample_bioseqfile.py
RefactoringTool: Files that need to be modified:
RefactoringTool: riboviz/subsample_bioseqfile.py

# riboviz/tools/subsample_bioseqfile.py: 
(riboviz) fanders6@ceres:~/riboviz/riboviz$ 2to3 riboviz/tools/subsample_bioseqfile.py 
RefactoringTool: Skipping optional fixer: buffer
RefactoringTool: Skipping optional fixer: idioms
RefactoringTool: Skipping optional fixer: set_literal
RefactoringTool: Skipping optional fixer: ws_comma
RefactoringTool: No changes to riboviz/tools/subsample_bioseqfile.py
RefactoringTool: Files that need to be modified:
RefactoringTool: riboviz/tools/subsample_bioseqfile.py
```

Branch `subsample-fastq-38` is up to date with latest develop, and I think that this is ready to merge and close...  pending @mikej888 's opinions: 
* on whether pylint's "Too many local variables" is avoidable/ignorable, 
* and how/whether to fix pycodestyle's "E124 closing bracket does not match visual indentation" issue without causing further indent issues.  I didn't have much success trying to iron this one out, and I'm not sure whether pylint and pycodestyle have slightly differing opinions on where to put these brackets.

ewallace commented:
I request that the seed value is an optional input parameter, with default value `seedvalue = Null` (or pythonic equivalent). I don't know if `seedvalue` is the right name, but `seed` is already a function name.

Then (pseudocode):
```
if !is.null(seedvalue) :
    random.seed(seedvalue)
```

It's ok to use `seedvlaue = 42` for a test.

The point is that we don't want to hard-code a specific random number generator. We want the option to fix a seed for reproducibility, but to allow users to generate different random subsets, for example for cross-validation analysis.

mikej888 commented:
@FlicAnderson 

> on whether pylint's "Too many local variables" is avoidable/ignorable,

You can avoid it by extracting row_count, row_count_out and with block into a new function but I don't think that gains anything. Personally I'd ignore it. The style checker is a goal to strive for not a hard and fast rule. For example the `pandas.DataFrame.to_csv` function has over 20 arguments which is well beyond the 5 that pylint hopes for!

> and how/whether to fix pycodestyle's "E124 closing bracket does not match visual indentation" issue without causing further indent issues. I didn't have much success trying to iron this one out, and I'm not sure whether pylint and pycodestyle have slightly differing opinions on where to put these brackets. 

I had a guddle and could only get that to pass by changing

```
    print((("subsampling complete; read {} records from {}, wrote {} records \
to {}".format(row_count, seqfilein, row_count_out, seqfileout))
          ))
```
to:
```
    print((("subsampling complete; read {} records from {}, wrote {} records \
to {}".format(row_count, seqfilein, row_count_out, seqfileout))))
```
I think in this case nothing is gained by having the closing brackets on their own line.
One set of brackets is redundant and could be removed:
```
    print(("subsampling complete; read {} records from {}, wrote {} records \
to {}".format(row_count, seqfilein, row_count_out, seqfileout)))
```

FlicAnderson commented:
As of commit: d77fb2c

Updated code to add `seedvalue` parameter as per @ewallace's [suggestion](https://github.com/riboviz/riboviz/pull/234#issuecomment-776954932), and used @mikej888 's [fix](https://github.com/riboviz/riboviz/pull/234#issuecomment-777361486) for the "E124 closing bracket does not match visual indentation" issue from pycodestyle. 

Pylint still returning the "too many arguments" and "too many local variables" issues, but as Mike says, it's ok to proceed ignoring those. 

```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/subsample_bioseqfile.py 
************* Module riboviz.subsample_bioseqfile
riboviz/subsample_bioseqfile.py:10:0: R0913: Too many arguments (7/5) (too-many-arguments)
riboviz/subsample_bioseqfile.py:10:0: R0914: Too many local variables (17/15) (too-many-locals)
------------------------------------------------------------------
Your code has been rated at 9.39/10 (previous run: 6.36/10, +3.03)

(riboviz) fanders6@ceres:~/riboviz/riboviz$ pylint riboviz/tools/subsample_bioseqfile.py 
-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.15/10, +1.85)
```
```
# file exists; run WITHOUT overwrite parameter
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.001 -s 42
Created by: RiboViz
Date: 2021-02-23 20:15:16.851022
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit c18d016dcdc7d3c23bfbcb6dae4aee1087904740 date 2021-02-10 18:03:01+00:00

Traceback (most recent call last):
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/fanders6/miniconda3/envs/riboviz/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 117, in <module>
    invoke_subsample_bioseqfile()
  File "/home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py", line 113, in invoke_subsample_bioseqfile
    verbose)
  File "/home/fanders6/riboviz/riboviz/riboviz/subsample_bioseqfile.py", line 43, in subsample_bioseqfile
    .format(seqfileout))
ValueError: output file vignette/tmp/subsample.fastq.gz already exists, use '-overwrite' to replace

# WITH overwrite parameter:
(riboviz) fanders6@ceres:~/riboviz/riboviz$ python -m riboviz.tools.subsample_bioseqfile -i vignette/input/SRR1042855_s1mi.fastq.gz -o vignette/tmp/subsample.fastq.gz -p 0.001 -s 42 -f True
Created by: RiboViz
Date: 2021-02-23 20:15:59.384828
Command-line tool: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
File: /home/fanders6/riboviz/riboviz/riboviz/tools/subsample_bioseqfile.py
Version: commit c18d016dcdc7d3c23bfbcb6dae4aee1087904740 date 2021-02-10 18:03:01+00:00
read 100000
read 200000
read 300000
read 400000
read 500000
read 600000
read 700000
read 800000
read 900000
subsampling complete; read 963571 records from vignette/input/SRR1042855_s1mi.fastq.gz, wrote 921 records to vignette/tmp/subsample.fastq.gz
```

As I edit `seedvalue` (in the above command), the number of reads changes at different random seed values (e.g. 948 records at seed 1, 942 records at seed 35, 921 at seed 42), as expected.

Not sure whether this requires a final review since it's been mostly reviewed a few times.
